### PR TITLE
Add connection state watcher

### PR DESCRIPTION
**Pull request by Jarema**
Signed-off-by: Tomasz Pietrek <tomasz@nats.io>

### CODE REVIEW

File: async-nats/tests/client_tests.rs [hunk 1]
```diff
@@ -562,4 +563,14 @@ mod client {
...
     }
+
+    #[tokio::test]
+    async fn connection_state() {
+        let server = nats_server::run_basic_server();
+        let client = async_nats::connect(server.client_url()).await.unwrap();
+        assert_eq!(State::Connected, client.connection_state());
+        drop(server);
+        tokio::time::sleep(Duration::from_secs(1)).await;
+        assert_eq!(State::Disconnected, client.connection_state());
```
Comment (caspervonb) [inline async-nats/tests/client_tests.rs:574]: Test for reconnect too?

File: async-nats/src/connection.rs [hunk 2]
```diff
@@ -30,6 +31,25 @@ pub(crate) trait AsyncReadWrite: AsyncWrite + AsyncRead + Send + Unpin {}
 /// Blanked implementation that applies to both TLS and non-TLS `TcpStream`.
 impl<T> AsyncReadWrite for T where T: AsyncRead + AsyncWrite + Unpin + Send {}
 
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub enum State {
+    Pending,
+    Connected,
+    Disconnected,
+    Closed,
```
Comment (caspervonb) [inline async-nats/src/connection.rs:39]: Currently, we can't actually reach closed can we?

```suggestion
    Closed,
```

File: async-nats/tests/client_tests.rs (hunk 1 above)
Comment (Jarema) [reply async-nats/tests/client_tests.rs:575]: hard to do here. Added reconnect check in reconnection general test.

**Review by caspervonb — APPROVED** [review]
lgtm

**Pull request merged**